internal/bufpushaction/bufpushaction_test.go

```diff
@@ -0,0 +1,322 @@
+		pushTest{
+			track: testNonMainTrack,
+			bufRuns: []fakeCommandRunnerRun{
+				trackHelpRun(testNonMainTrack, true),
+				{
+					expectArgs: []string{"beta", "registry", "commit", "get", testModuleNonMainTrack, "--format", "json"},
+					stderr:     fmt.Sprintf("Failure: %q does not exist", testModuleName),
+					err:        assert.AnError,
+				},
+				{
+					expectArgs: []string{"push", "--track", testNonMainTrack, "--tag", testGitCommit1, testInput},
+					stderr:     repoNotFoundMessage,
+					err:        assert.AnError,
+				},
+			},
+			errorAssertion: func(err error) {
+				require.EqualError(t, err, repoNotFoundMessage)
+			},
+		}.test(t)
+	})
+
+	t.Run("push commit with same digest as head", func(t *testing.T) {
+		dupContentMessage := "The latest commit has the same content; not creating a new commit."
+		pushTest{
+			track: testNonMainTrack,
+			bufRuns: []fakeCommandRunnerRun{
+				trackHelpRun(testNonMainTrack, true),
+				getTagsRun(t, testModuleNonMainTrack, testGitCommit2),
+				{
+					expectArgs: []string{"push", "--track", testNonMainTrack, "--tag", testGitCommit1, testInput},
+					stderr:     dupContentMessage,
+				},
+				{
+					expectArgs: []string{"beta", "registry", "commit", "get", testModuleNonMainTrack, "--format", "json"},
+					stdout:     buildCommitJSON(t, testBsrCommit),
+				},
+			},
+			compareCommitRuns: []fakeCompareCommits{
+				compareCommitsRun(testGitCommit2, testGitCommit1, github.CompareCommitsStatusAhead),
+			},
+			expectStdout: []string{
+				workflowNotice(dupContentMessage),
+				workflowOutput("commit", testBsrCommit),
+				workflowOutput("commit_url", bsrCommitURL(testModuleName, testBsrCommit)),
+			},
+		}.test(t)
+	})
+
+	t.Run("tagged git commit not found on github", func(t *testing.T) {
+		goGithubNotFoundError := &gogithub.ErrorResponse{
+			Response: &http.Response{
+				StatusCode: http.StatusNotFound,
+				Request:    &http.Request{},
+			},
+		}
+		pushTest{
+			track: testNonMainTrack,
+			bufRuns: []fakeCommandRunnerRun{
+				trackHelpRun(testNonMainTrack, true),
+				getTagsRun(t, testModuleNonMainTrack, testGitCommit2),
+				{
+					expectArgs: []string{"push", "--track", testNonMainTrack, "--tag", testGitCommit1, testInput},
+					stdout:     testBsrCommit,
+				},
+			},
+			compareCommitRuns: []fakeCompareCommits{
+				{
+					expectBase: testGitCommit2,
+					expectHead: testGitCommit1,
+					err:        goGithubNotFoundError,
+				},
+			},
+			expectStdout: []string{
+				workflowOutput("commit", testBsrCommit),
+				workflowOutput("commit_url", bsrCommitURL(testModuleName, testBsrCommit)),
+			},
+		}.test(t)
+	})
+}
+
+type pushTest struct {
+	track             string
+	bufRuns           []fakeCommandRunnerRun
+	compareCommitRuns []fakeCompareCommits
+	expectStdout      []string
+	expectStderr      []string
+	errorAssertion    func(err error)
+}
+
+func (pt pushTest) test(t *testing.T) {
```

Comment:
I find the way these tests are written a bit confusing. The structure here very much resembles a table test that takes a struct as input and runs the same logic on all, but by using a method on the param struct? Is that a pattern we started applying in bufbuild/buf?

Reply:
I swapped the test method out for a function.